crypt-yescrypt.c

```diff
@@ -39,71 +50,101 @@ gensalt_yescrypt_rn(unsigned long count,
-		output[1] = '\0';
-		errno = ERANGE;
-		return;
-	}
+  if (o_size < 3 + 8 + 1 + BASE64_LEN (nrbytes) ||
+      CRYPT_GENSALT_OUTPUT_SIZE < 3 + 8 + 1 + BASE64_LEN (nrbytes))
+    {
+      errno = ERANGE;
+      return;
+    }
+
+  if (count > 11)
+    {
+      errno = EINVAL;
+      return;
+    }
+
+  /* Temporary buffer for operation.  The buffer is guaranteed to be
+     large enough to hold the maximum size of the generated salt and
+     is filled with '\0' bytes before use.  */
+  uint8_t outbuf[CRYPT_GENSALT_OUTPUT_SIZE];
+  XCRYPT_SECURE_MEMSET(outbuf, CRYPT_GENSALT_OUTPUT_SIZE);
+
+  yescrypt_params_t params =
+  {
+    .flags = YESCRYPT_DEFAULTS,
+    .p = 1,
+  };
+
+  /* Valid cost parameters are from 1 to 11.  The default is 5.
+     These are used to set yescrypt's 'N' and 'r' parameters as
+     follows:
+     N (block count) is specified in units of r (block size).
+     The author of yescrypt recommends in the documentation to use
+     r=8 for total sizes of 2 MiB and less, and r=32 above that.
+     This has to do with the typical per-core last-level cache sizes
+     of current CPUs.  */
+
+  if (count == 0)
+    count = 5;
+
+  if (count < 3)
+    {
+      params.r = 8;
+      params.N = 1ULL << (count + 9); // 1 -> 1024, 2 -> 2048
+    }
+  else
+    {
+      params.r = 32;
+      params.N = 1ULL << (count + 7); // 3 -> 1024, 4 -> 2048, ... 11 -> 262144
+    }
+
+  if (!yescrypt_encode_params_r(&params, rbytes, nrbytes, outbuf, o_size))
+    {
+      errno = ERANGE;
+      return;
+    }
+
+  memcpy(output, outbuf, o_size);
+  return;
 }
 
 void
-crypt_yescrypt_rn(const char *phrase, size_t ARG_UNUSED (phr_size),
-                  const char *setting, size_t ARG_UNUSED (set_size),
+crypt_yescrypt_rn(const char *phrase, size_t phr_size,
+                  const char *setting, size_t set_size,
                   uint8_t *output, size_t o_size,
-                  ARG_UNUSED(void *scratch), ARG_UNUSED(size_t s_size))
+                  void *scratch, size_t s_size)
 {
-	yescrypt_local_t local;
-	uint8_t *retval;
-
-	if (o_size < 3) {
-		errno = ERANGE;
-		return;
-	}
-	if (yescrypt_init_local(&local)) {
-		errno = ENOMEM;
-		return;
-	}
-	retval = yescrypt_r(NULL, &local,
-	    (const uint8_t *)phrase, strlen(phrase),
-	    (const uint8_t *)setting, NULL,
-	    output, o_size);
-	if (yescrypt_free_local(&local) ||
-	    !retval) {
-		/*
-		 * As the output could have already been written,
-		 * overwrite it with a failure token.
-		 */
-		output[0] = '*';
-		output[1] = '0';
-		output[2] = '\0';
-		errno = EINVAL;
-	}
+  if (o_size < set_size + 1 + 43 + 1 ||
+      CRYPT_OUTPUT_SIZE < set_size + 1 + 43 + 1 ||
+      s_size < sizeof(crypt_yescrypt_internal_t))
+    {
+      errno = ERANGE;
+      return;
+    }
+
+  crypt_yescrypt_internal_t *intbuf = scratch;
+  XCRYPT_SECURE_MEMSET(intbuf->outbuf, CRYPT_OUTPUT_SIZE);
+
+  if (yescrypt_init_local (&intbuf->local))
+    {
+      if (errno == 0)
+        errno = ENOMEM;
```

Comment:
I think proposed change is invalid. As we should never rely on reading of `errno`, which is never set or reinitialized to zero. And it would be non-deterministic behavior for `libcrypt` user if function result is dependent on reading of uninitialized value (for example when called before or after of some syscall which changes `errno`). I think, as @zackw suggested and @solardiz agreed this call should "just return".

Reply:
The wording in C11 relating to the initial value of `errno` leaves something to be desired, but I could argue that it's a conformance violation if it doesn't _start out_ as zero upon entry to `main`. However, I agree that this makes the behavior nondeterministic based on what has happened earlier in the program, and we should "just return" instead.